gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```
@@ -0,0 +1,191 @@
+
+  private String metadataPath;
+  private EnvironmentProvider envProvider;
+  private ProcessProvider processProvider;
+
+  public enum UseMtlsEndpoint {
+    NEVER,
+    AUTO,
+    ALWAYS;
+  }
+
+  @VisibleForTesting
+  MtlsProvider(
+      EnvironmentProvider envProvider, ProcessProvider processProvider, String metadataPath) {
+    this.envProvider = envProvider;
+    this.processProvider = processProvider;
+    this.metadataPath = metadataPath;
+  }
+
+  public MtlsProvider() {
+    this(
+        new SystemEnvironmentProvider(),
+        new DefaultProcessProvider(),
+        DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+  }
+
+  /**
+   * Returns if mutual TLS client certificate should be used. If the value is true, the key store
+   * from {@link #getKeyStore()} will be used to configure mutual TLS transport.
+   */
+  public boolean useMtlsClientCertificate() {
+    String useClientCertificate = envProvider.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE");
+    return "true".equals(useClientCertificate);
+  }
+
+  /**
+   * Returns AUTO, NEVER or ALWAYS. NEVER means always use regular endpoint; ALWAYS means always use
+   * mTLS endpoint; AUTO means auto switch to mTLS endpoint if client certificate exists and should
+   * be used.
+   */
+  public UseMtlsEndpoint useMtlsEndpoint() {
+    String useMtlsEndpoint = envProvider.getenv("GOOGLE_API_USE_MTLS_ENDPOINT");
+    if ("never".equals(useMtlsEndpoint)) {
+      return UseMtlsEndpoint.NEVER;
+    } else if ("always".equals(useMtlsEndpoint)) {
+      return UseMtlsEndpoint.ALWAYS;
+    }
+    return UseMtlsEndpoint.AUTO;
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (FileNotFoundException exception) {
+      // file doesn't exist
```

Do we need to propagate an error (e.g. via an exception)?

---

The metadata file is a well known path, if a machine doesn't have endpoint verification support, then it doesn't have this file, so we just return null. This is not an error.

---

SG. This state is non-obvious, and the comment helps clear that up - could we add it to the source?

---

updated the comments